contracts/src/v0.8/automation/interfaces/IAutomationV21PlusCommon.sol

```diff
@@ -0,0 +1,212 @@
+   * @member amountSpent the amount this upkeep has spent
+   * @member paused if this upkeep has been paused
+   * @member offchainConfig the off-chain config of this upkeep
+   */
+  struct UpkeepInfo {
+    address target;
+    uint32 performGas;
+    bytes checkData;
+    uint96 balance;
+    address admin;
+    uint64 maxValidBlocknumber;
+    uint32 lastPerformedBlockNumber;
+    uint96 amountSpent;
+    bool paused;
+    bytes offchainConfig;
+  }
+
+  /**
+   * @notice the trigger structure conditional trigger type
+   */
+  struct ConditionalTrigger {
+    uint32 blockNum;
+    bytes32 blockHash;
+  }
+
+  /**
+   * @notice the trigger structure of log upkeeps
+   * @dev NOTE that blockNum / blockHash describe the block used for the callback,
+   * not necessarily the block number that the log was emitted in!!!!
+   */
+  struct LogTrigger {
+    bytes32 logBlockHash;
+    bytes32 txHash;
+    uint32 logIndex;
+    uint32 blockNum;
+    bytes32 blockHash;
+  }
+
+  function checkUpkeep(
+    uint256 id,
+    bytes memory triggerData
+  )
+    external
+    view
+    returns (
+      bool upkeepNeeded,
+      bytes memory performData,
+      uint8 upkeepFailureReason,
+      uint256 gasUsed,
+      uint256 gasLimit,
+      uint256 fastGasWei,
+      uint256 linkNative
+    );
+  function checkUpkeep(
+    uint256 id
+  )
+    external
+    view
+    returns (
+      bool upkeepNeeded,
+      bytes memory performData,
+      uint8 upkeepFailureReason,
+      uint256 gasUsed,
+      uint256 gasLimit,
+      uint256 fastGasWei,
+      uint256 linkNative
+    );
+  function simulatePerformUpkeep(
+    uint256 id,
+    bytes memory performData
+  ) external view returns (bool success, uint256 gasUsed);
+  function executeCallback(
+    uint256 id,
+    bytes memory payload
+  ) external returns (bool upkeepNeeded, bytes memory performData, uint8 upkeepFailureReason, uint256 gasUsed);
+  function checkCallback(
+    uint256 id,
+    bytes[] memory values,
+    bytes memory extraData
+  ) external view returns (bool upkeepNeeded, bytes memory performData, uint8 upkeepFailureReason, uint256 gasUsed);
+  function typeAndVersion() external view returns (string memory);
+  function addFunds(uint256 id, uint96 amount) external;
+  function cancelUpkeep(uint256 id) external;
+
+  function getUpkeepPrivilegeConfig(uint256 upkeepId) external view returns (bytes memory);
+  function hasDedupKey(bytes32 dedupKey) external view returns (bool);
+  function getUpkeepTriggerConfig(uint256 upkeepId) external view returns (bytes memory);
+  function getUpkeep(uint256 id) external view returns (UpkeepInfo memory upkeepInfo);
+  function getMinBalance(uint256 id) external view returns (uint96);
+  function getState()
+    external
+    view
+    returns (
+      AutomationRegistryBase2_2.State memory state,
+      AutomationRegistryBase2_2.OnchainConfigLegacy memory config,
```

Comment:
These are not really 2.2 structs, we want to repurpose them as 2.1+ structs. So i would say we should define this anew in this file and import in relevant versions

Reply:
Also let's be consistent with naming (Legacy.,non-legacy). What does legacy mean in this context?

Reply:
+1 to defining these structs in this interface. and then importing them in our 2.1, 2.2, and 2.3 implementations

Reply:
great. i define shared structs in this interface and replace their counterparts in 2.1 and 2.2. 
i will leave 2.3 out of this PR and Lei or Ryan and use this interface to adjust 2.3